# **Nombre: Mateo Orozco Baldovino**

In [23]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [41]:
# Importar archivo salud

df = pd.read_excel('Salud.xlsx')
np.random.seed(1)
df.head()

ID  exerany
0   1        0
1   2        1
2   3        1
3   4        1
4   5        0

In [25]:
print(f"Número total de personas en el dataset: {len(df)}")

Número total de personas en el dataset: 1000


In [26]:
N = 1000


In [27]:
# Parámetros del muestreo sistemático
n = 100  # tamaño de la muestra deseada
k = N // n  # intervalo de muestreo
r = np.random.randint(1, k+1)  # punto de inicio aleatorio



In [47]:
# Asumiendo que N (tamaño de la población) ya está definido, por ejemplo:
N = 1000 # Ejemplo: tamaño de tu DataFrame df
# Parámetros del muestreo sistemático
n = 100  # tamaño de la muestra deseada
k = N // n  # intervalo de muestreo
np.random.seed(6)
r = np.random.randint(1, k+1)

In [48]:
# Selección sistemática
indices_muestra = np.arange(r-1, N, k)  # -1 porque Python indexa desde 0
muestra = df.iloc[indices_muestra]

print(f"Tamaño de la población: {N}")
print(f"Tamaño de la muestra: {n}")
print(f"Intervalo de muestreo: {k}")
print(f"Punto de inicio aleatorio: {r}")


Tamaño de la población: 1000
Tamaño de la muestra: 100
Intervalo de muestreo: 10
Punto de inicio aleatorio: 10


In [29]:
variable_interes = 'exerany'

In [30]:
# Asegurarse de que la columna 'exerany' sea numérica (0 o 1)
df[variable_interes] = pd.to_numeric(df[variable_interes], errors='coerce')
df = df.dropna(subset=[variable_interes]) # Eliminar filas con NaN si hubo errores en la conversión
df[variable_interes] = df[variable_interes].astype(int) # Asegurar que sean enteros 0 o 1

In [31]:
# Proporción de personas que hicieron ejercicio para la población completa (referencia)
proporcion_poblacion = df[variable_interes].mean()
print(f"\nProporción en la población completa (referencia): {proporcion_poblacion:.4f}")


Proporción en la población completa (referencia): 0.6560


In [34]:
# Proporción en la muestra Sistemática
proporcion_sistematica = muestra[variable_interes].mean()
print(f"Proporción de ejercicio en la Muestra Sistemática: {proporcion_sistematica:.4f}")

Proporción de ejercicio en la Muestra Sistemática: 0.6300


In [35]:
n = 100
muestra_mas = df.sample(n=n, replace=False)
proporcion_mas = muestra_mas[variable_interes].mean()

In [36]:
# Proporción en la muestra MAS
proporcion_mas = muestra_mas[variable_interes].mean()
print(f"Proporción de ejercicio en la Muestra Aleatoria Simple (MAS): {proporcion_mas:.4f}")

Proporción de ejercicio en la Muestra Aleatoria Simple (MAS): 0.6800


In [61]:
import pandas as pd
import numpy as np
# Función para realizar muestreo MAS y sistemático y calcular la proporción
def muestreo_y_proporcion(df, n, variable_interes, seed):
  """
  Realiza muestreo MAS y sistemático y calcula la proporción de una variable.

  Args:
    df: DataFrame de entrada.
    n: Tamaño de la muestra.
    variable_interes: Nombre de la columna de interés (binaria 0/1).
    seed: Semilla para la reproducibilidad de los resultados.

  Returns:
    Un diccionario con las proporciones para MAS y Sistemático.
  """
  N = len(df)
  k = N // n

  # Muestreo Aleatorio Simple (MAS)
  np.random.seed(seed)
  muestra_mas = df.sample(n=n, replace=False)
  proporcion_mas = muestra_mas[variable_interes].mean()

  # Muestreo Sistemático
  np.random.seed(seed)
  r = np.random.randint(1, k + 1)
  indices_muestra = np.arange(r - 1, N, k)
  muestra_sistematica = df.iloc[indices_muestra]
  proporcion_sistematica = muestra_sistematica[variable_interes].mean()

  return {
      'seed': seed,
      'proporcion_mas': proporcion_mas,
      'proporcion_sistematica': proporcion_sistematica,
      'punto_inicio_sistematico': r
  }

# Repetir el proceso 5 veces con diferentes semillas
n_repeticiones = 5
resultados = []
variable_interes = 'exerany' # Asegúrate de que esta columna exista y esté preprocesada como 0/1

# Asegurarse de que la columna 'exerany' sea numérica (0 o 1) y sin NaNs
df[variable_interes] = pd.to_numeric(df[variable_interes], errors='coerce')
df = df.dropna(subset=[variable_interes])
df[variable_interes] = df[variable_interes].astype(int)

for i in range(n_repeticiones):
  seed = i + 1 # Usar semillas del 1 al 5
  resultado = muestreo_y_proporcion(df.copy(), n, variable_interes, seed) # Usar una copia para evitar modificar el df original
  resultados.append(resultado)

# Imprimir los resultados
print("Resultados de los muestreos y proporciones:")
for res in resultados:
  print(f"\n--- Semilla: {res['seed']} ---")
  print(f"Muestreo Aleatorio Simple (MAS): Proporción = {res['proporcion_mas']:.4f}")
  print(f"Muestreo Sistemático: Proporción = {res['proporcion_sistematica']:.4f} (Punto de inicio: {res['punto_inicio_sistematico']})")



Resultados de los muestreos y proporciones:

--- Semilla: 1 ---
Muestreo Aleatorio Simple (MAS): Proporción = 0.5900
Muestreo Sistemático: Proporción = 0.6500 (Punto de inicio: 6)

--- Semilla: 2 ---
Muestreo Aleatorio Simple (MAS): Proporción = 0.5800
Muestreo Sistemático: Proporción = 0.6300 (Punto de inicio: 9)

--- Semilla: 3 ---
Muestreo Aleatorio Simple (MAS): Proporción = 0.6600
Muestreo Sistemático: Proporción = 0.6300 (Punto de inicio: 9)

--- Semilla: 4 ---
Muestreo Aleatorio Simple (MAS): Proporción = 0.5800
Muestreo Sistemático: Proporción = 0.6400 (Punto de inicio: 8)

--- Semilla: 5 ---
Muestreo Aleatorio Simple (MAS): Proporción = 0.6900
Muestreo Sistemático: Proporción = 0.6300 (Punto de inicio: 4)


**¿Cuál diseño produce estimaciones más estables?**

El Muestreo Aleatorio Simple (MAS) parece producir estimaciones ligeramente menos estables en este conjunto de pruebas. Sus proporciones varían desde 0.5800 hasta 0.6900. Por otro lado, el Muestreo Sistemático muestra una variación más contenida, con proporciones que oscilan entre 0.6300 y 0.6500 (incluyendo el punto de inicio que podría influir), sugiriendo una mayor estabilidad en sus estimaciones a través de diferentes semillas y puntos de inicio.

**¿Observas diferencias en la variabilidad?**

Sí, se observan diferencias en la variabilidad entre ambos diseños. La variabilidad de las proporciones obtenidas por Muestreo Aleatorio Simple (MAS) es mayor, cubriendo un rango más amplio de valores (0.5800 a 0.6900). En contraste, el Muestreo Sistemático presenta una menor variabilidad en las proporciones (0.6300 a 0.6500), lo que indica que sus estimaciones tienden a agruparse más estrechamente.

**¿Qué limitaciones tiene el muestreo sistemático en este caso?**

En este caso, la principal limitación del muestreo sistemático radica en la posibilidad de sesgo si existe un patrón o periodicidad subyacente en la lista ordenada de la población que coincide con el intervalo de muestreo (k). Si, por ejemplo, cada décimo ID en el dataset de 'Salud' correspondiera a un tipo específico de persona con una tendencia particular a hacer ejercicio, el muestreo sistemático podría subestimar o sobrestimar consistentemente la proporción real. Además, la calidad de la muestra sistemática depende en gran medida del punto de inicio aleatorio (r), ya que un mal punto de inicio en una lista con patrones podría sesgar la muestra.

**¿Cuál diseño recomendarías y por qué?**

Para la mayoría de los escenarios generales donde no se tiene conocimiento previo de patrones en la población o una ordenación particular que pudiera introducir sesgos, recomendaría el Muestreo Aleatorio Simple (MAS). Aunque en esta simulación particular el sistemático mostró un poco más de estabilidad, el MAS es metodológicamente más robusto y tiene una base teórica sólida (como la que fundamenta el Teorema Central del Límite) que garantiza que, en promedio, sus estimaciones serán imparciales y su variabilidad puede ser cuantificada con precisión. Sin embargo, si la población está naturalmente ordenada de una manera que no introduce un sesgo periódico y la implementación logística es más sencilla (por ejemplo, muestreo de clientes en una línea de producción), el muestreo sistemático podría ser una alternativa eficiente.

In [91]:
# Importar dataframe pasajeros

df_pasajeros = pd.read_excel('pasajeros.xlsx')
df_pasajeros.head()

# Variable de interés: Pasajeros por mes (2015-2024)
variable_interes2 = 'Pasajeros'

# Asegúrate de que la columna 'Fecha' sea de tipo datetime si aún no lo es
df_pasajeros['Fecha'] = pd.to_datetime(df_pasajeros['Fecha'])

# Establecer la columna 'Fecha' como índice para facilitar el muestreo basado en tiempo
df_pasajeros = df_pasajeros.set_index('Fecha')

# Calcular el número total de meses disponibles
N_meses = len(df_pasajeros)

# Tamaño de la muestra deseada (n=12 meses)
n_muestra_meses = 12

# --- Muestreo Aleatorio Simple (MAS) de 12 meses ---
np.random.seed(6) # para reproducibilidad
indices_aleatorios = np.random.choice(N_meses, size=n_muestra_meses, replace=False)
muestra_mas_meses = df_pasajeros.iloc[indices_aleatorios]


# --- Muestreo Sistemático de 12 meses ---
# Calcular el intervalo de muestreo
k_meses = N_meses // n_muestra_meses

# Seleccionar un punto de inicio aleatorio
np.random.seed(42) # Usar la misma semilla o una diferente si se desea
r_meses = np.random.randint(1, k_meses + 1)

# Seleccionar los índices para el muestreo sistemático
indices_sistematico_meses = np.arange(r_meses - 1, N_meses, k_meses)


if len(indices_sistematico_meses) < n_muestra_meses:

    indices_sistematico_meses = np.arange(r_meses - 1, N_meses, k_meses)[:n_muestra_meses]


muestra_sistematica_meses = df_pasajeros.iloc[indices_sistematico_meses]

print("\n--- Muestra Sistemática de 12 meses ---")
print(f"Intervalo de muestreo (k): {k_meses}")
print(f"Punto de inicio aleatorio (r): {r_meses}")


# --- Calcular la media mensual de pasajeros ---
# Media para la muestra MAS
media_mas_pasajeros = muestra_mas_meses[variable_interes2].mean()
print(f"\nMedia mensual de pasajeros en la Muestra Aleatoria Simple (MAS): {media_mas_pasajeros:.2f}")

# Media para la muestra Sistemática
media_sistematica_pasajeros = muestra_sistematica_meses[variable_interes2].mean()
print(f"Media mensual de pasajeros en la Muestra Sistemática: {media_sistematica_pasajeros:.2f}")

# Media para la población completa (opcional, como referencia)
media_poblacion_pasajeros = df_pasajeros[variable_interes2].mean()
print(f"Media mensual de pasajeros en la Población completa: {media_poblacion_pasajeros:.2f}")




--- Muestra Sistemática de 12 meses ---
Intervalo de muestreo (k): 10
Punto de inicio aleatorio (r): 7

Media mensual de pasajeros en la Muestra Aleatoria Simple (MAS): 142411.50
Media mensual de pasajeros en la Muestra Sistemática: 151844.33
Media mensual de pasajeros en la Población completa: 149603.67


In [92]:
import pandas as pd
import numpy as np
# Función para realizar muestreo MAS y sistemático y calcular la media
def muestreo_y_media(df, n, variable_interes, seed):

  N = len(df)
  k = N // n

  # Muestreo Aleatorio Simple (MAS)
  np.random.seed(seed)
  indices_aleatorios = np.random.choice(N, size=n, replace=False)
  muestra_mas = df.iloc[indices_aleatorios]
  media_mas = muestra_mas[variable_interes].mean()

  # Muestreo Sistemático
  np.random.seed(seed)
  r = np.random.randint(1, k + 1)
  indices_sistematico = np.arange(r - 1, N, k)

  # Ajustar si el último paso excede el tamaño deseado
  if len(indices_sistematico) > n:
      indices_sistematico = indices_sistematico[:n]
  elif len(indices_sistematico) < n:
      # Esto puede ocurrir si N % n != 0 y el cálculo inicial de k es un entero truncado
      # Una solución es ajustar k ligeramente o permitir un tamaño de muestra ligeramente diferente
      # Para mantener n fijo, podemos intentar incluir el último elemento si n no se alcanza
      if N not in indices_sistematico and len(indices_sistematico) < n:
          indices_sistematico = np.append(indices_sistematico, N-1)
      # Si aún no se alcanza n, se podría necesitar un ajuste más complejo o aceptar un tamaño ligeramente menor
      indices_sistematico = indices_sistematico[:n] # Asegurar que no exceda n, aunque podría ser menor

  muestra_sistematica = df.iloc[indices_sistematico]
  media_sistematica = muestra_sistematica[variable_interes].mean()


  return {
      'seed': seed,
      'media_mas': media_mas,
      'media_sistematica': media_sistematica,
      'punto_inicio_sistematico': r,
      'tamano_sistematico_real': len(muestra_sistematica) # Opcional: para verificar el tamaño real
  }

# Repetir el proceso 5 veces con diferentes semillas para el dataset de pasajeros
n_repeticiones = 5
resultados_pasajeros = []
variable_interes2 = 'Pasajeros' # Asegúrate de que esta columna exista y sea numérica
n_muestra_meses = 12

# Asegurarse de que la columna de interés sea numérica
df_pasajeros[variable_interes2] = pd.to_numeric(df_pasajeros[variable_interes2], errors='coerce')
df_pasajeros = df_pasajeros.dropna(subset=[variable_interes2]) # Eliminar NaNs si hay

for i in range(n_repeticiones):
  seed = i + 1 # Usar semillas diferentes, por ejemplo del 1 al 5
  resultado = muestreo_y_media(df_pasajeros.copy(), n_muestra_meses, variable_interes2, seed) # Usar una copia
  resultados_pasajeros.append(resultado)

# Imprimir los resultados para el dataset de pasajeros
print("\nResultados de los muestreos y medias (Dataset Pasajeros):")
for res in resultados_pasajeros:
  print(f"\n--- Semilla: {res['seed']} ---")
  print(f"Muestreo Aleatorio Simple (MAS): Media = {res['media_mas']:.2f}")
  print(f"Muestreo Sistemático: Media = {res['media_sistematica']:.2f} (Punto de inicio: {res['punto_inicio_sistematico']}, Tamaño real: {res['tamano_sistematico_real']})")

# Media para la población completa (referencia)
media_poblacion_pasajeros = df_pasajeros[variable_interes2].mean()
print(f"\nMedia mensual de pasajeros en la Población completa (referencia): {media_poblacion_pasajeros:.2f}")


Resultados de los muestreos y medias (Dataset Pasajeros):

--- Semilla: 1 ---
Muestreo Aleatorio Simple (MAS): Media = 163359.50
Muestreo Sistemático: Media = 150096.17 (Punto de inicio: 6, Tamaño real: 12)

--- Semilla: 2 ---
Muestreo Aleatorio Simple (MAS): Media = 146625.67
Muestreo Sistemático: Media = 152398.08 (Punto de inicio: 9, Tamaño real: 12)

--- Semilla: 3 ---
Muestreo Aleatorio Simple (MAS): Media = 155874.00
Muestreo Sistemático: Media = 152398.08 (Punto de inicio: 9, Tamaño real: 12)

--- Semilla: 4 ---
Muestreo Aleatorio Simple (MAS): Media = 127016.08
Muestreo Sistemático: Media = 152327.67 (Punto de inicio: 8, Tamaño real: 12)

--- Semilla: 5 ---
Muestreo Aleatorio Simple (MAS): Media = 137796.42
Muestreo Sistemático: Media = 148164.58 (Punto de inicio: 4, Tamaño real: 12)

Media mensual de pasajeros en la Población completa (referencia): 149603.67


**¿Qué diferencias observas entre los métodos?**

El Muestreo Aleatorio Simple (MAS) muestra una mayor variabilidad en sus estimaciones de la media (desde 127016.08 hasta 163359.50) en comparación con el Muestreo Sistemático. El Muestreo Sistemático, por su parte, produce medias más consistentes y agrupadas (oscilando entre 148164.58 y 152398.08). Además, la "Media mensual de pasajeros en la Población completa" es 149603.67. El MAS (142411.50) subestimó la media real, mientras que el Sistemático (151844.33) estuvo más cerca, sugiriendo en este caso particular una mayor precisión del sistemático.

**¿Cuál capta mejor la estacionalidad o la tendencia?**

El Muestreo Sistemático es inherentemente más adecuado para captar la estacionalidad o la tendencia en series temporales (como datos mensuales de pasajeros) en comparación con el Muestreo Aleatorio Simple. Esto se debe a que el muestreo sistemático garantiza que los elementos se seleccionen a intervalos regulares a lo largo de la secuencia ordenada, lo que asegura que la muestra contenga una representación distribuida de diferentes puntos en el tiempo (meses, años) donde podrían manifestarse patrones estacionales o de tendencia. El MAS, al seleccionar elementos completamente al azar, podría, por pura casualidad, omitir ciertos periodos o concentrarse excesivamente en otros, perdiendo la representatividad de esos patrones cíclicos o evolutivos.

**¿Cuándo podría ser sesgado el muestreo sistemático?**

El muestreo sistemático podría ser seriamente sesgado si existe una periodicidad o un patrón cíclico en los datos de la población que coincide o es un múltiplo del intervalo de muestreo (k). Por ejemplo, si tu k fuera 12 (para datos mensuales) y hubiera un patrón anual muy marcado, podrías terminar seleccionando solo datos de enero de cada año (si tu punto de inicio cae en enero), o solo datos de junio, perdiendo la variación del resto del año y sesgando gravemente la estimación de la media anual o de la proporción. También podría ser sesgado si la lista no está verdaderamente aleatorizada al inicio (aunque en series temporales suele ser por orden cronológico).

**¿Qué método recomendarías para estudios con series temporales?**

Para estudios con series temporales donde se busca captar estacionalidad o tendencia, o simplemente asegurar una distribución equitativa de los datos a lo largo del tiempo, recomendaría el Muestreo Sistemático. Su naturaleza de selección a intervalos regulares es ventajosa para asegurar que la muestra contenga elementos de todo el período de estudio, reflejando así mejor los patrones temporales. Sin embargo, es crucial ser consciente de posibles patrones cíclicos en los datos que puedan coincidir con el intervalo de muestreo y, si se detectan, ajustar el diseño (ej., cambiando el punto de inicio aleatorio o usando un muestreo estratificado por períodos) para evitar el sesgo. Si no hay sospecha de dichos patrones, el sistemático es a menudo más práctico que el MAS para series largas.